Source/JavaScriptCore/builtins/WrapForValidIteratorPrototype.js

```diff
@@ -0,0 +1,79 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// https://tc39.es/proposal-iterator-helpers/#sec-wrapforvaliditeratorprototype.next
+function next()
+{
+    "use strict";
+
+    // 1. Let O be this value.
+    var o = this;
+    // 2. Perform ? RequireInternalSlot(O, [[Iterated]]).
+    // 3. Let iteratorRecord be O.[[Iterated]].
+    if (!@isObject(o))
+        @throwTypeError("baz1");
+    var iteratorRecord = @getByIdDirectPrivate(this, "iterated");
+    if (!@isObject(iteratorRecord))
+        @throwTypeError("baz2");
+    // 4. Return ? Call(iteratorRecord.[[NextMethod]], iteratorRecord.[[Iterator]]).
+    return iteratorRecord.nextMethod.@call(iteratorRecord.iterator);
+}
+
+// https://tc39.es/proposal-iterator-helpers/#sec-wrapforvaliditeratorprototype.return
+function return()
+{
+    "use strict";
+
+    // 1. Let O be this value.
+    var o = this;
+    // 2. Perform ? RequireInternalSlot(O, [[Iterated]]).
+    // 3. Let iteratorRecord be O.[[Iterated]].
+    if (!@isObject(o))
+        @throwTypeError("baz1");
+    var iteratorRecord = @getByIdDirectPrivate(this, "iterated");
+    if (!@isObject(iteratorRecord))
+        @throwTypeError("baz2");
+    // 3. Let iterator be O.[[Iterated]].[[Iterator]].
+    var iterator = iteratorRecord.iterator;
+    // 4. Assert: iterator is an Object.
+    @assert(@isObject(iterator));
+    // 5. Let returnMethod be ? GetMethod(iterator, "return").
+    var returnMethod = iterator.return;
+    // 6. If returnMethod is undefined, then
+    //   a. Return CreateIterResultObject(undefined, true).
+    if (returnMethod === @undefined)
+        return { value: @undefined, done: true };
+    // 7. Return ? Call(returnMethod, iterator).
+    return returnMethod.@call(iterator);
+}
+
+@linkTimeConstant
+@constructor
+function WrapForValidIterator(iteratorRecord)
+{
+    "use strict";
+
+    @putByIdDirectPrivate(this, "iterated", iteratorRecord);
```

Comment:
We would like to remove getDirect / putDirect things more and more. So let's implement it with JSInternalFieldObjectImpl instead.

Reply:
I've added `JSWrapForValidIterator` that extends `JSInternalFieldObjectImpl`